File: spawner.py
"""ZombieSpawner: positions & type selection for new zombies."""
from __future__ import annotations

import math
import random

import pygame

import settings as S


class ZombieSpawner:
    """Spawns zombies around the player (min distance enforced)."""

    def __init__(self) -> None:
        self.rng = random.Random()

    # -------------------------------------------------------------- types --
    def pick_type(self, wave: int, modifier: str = "none") -> str:
        """Weighted zombie type selection that unlocks as waves progress."""
        rng = self.rng
        weights: dict[str, float] = {"normal": 10.0}
        if wave >= 2:
            weights["fast"] = min(6.0, 2.0 + wave * 0.5)
        if wave >= 3:
            weights["tank"] = min(5.0, 1.0 + (wave - 2) * 0.4)
        if wave >= 4:
            weights["exploder"] = min(4.0, 1.0 + (wave - 3) * 0.35)
            weights["ranged"] = min(4.0, 1.0 + (wave - 3) * 0.35)
        if wave >= 5:
            weights["crawler"] = min(3.0, 1.0 + (wave - 4) * 0.25)
        if wave >= 7:
            weights["necromancer"] = min(2.0, 0.5 + (wave - 6) * 0.15)
        if modifier == "blood_moon":
            for k in weights:
                weights[k] *= 1.25
            weights["tank"] *= 1.5
        if modifier == "swarm":
            weights["crawler"] = weights.get("crawler", 0) * 2.0
            weights["fast"] *= 1.4
        kinds = list(weights)
        total = sum(weights.values())
        roll = rng.uniform(0, total)
        acc = 0.0
        for k in kinds:
            acc += weights[k]
            if roll <= acc:
                return k
        return "normal"

    # ----------------------------------------------------------- position --
    def spawn_position(self, player_pos: pygame.Vector2,
                       world_map, radius: float = 24.0) -> pygame.Vector2 | None:
        """Ring spawn around player: >= SPAWN_MIN_DIST away, free of walls."""
        for _ in range(40):
            ang = self.rng.uniform(0, math.tau)
            dist = self.rng.uniform(S.SPAWN_MIN_DIST, S.SPAWN_MAX_DIST)
            pos = player_pos + pygame.Vector2(math.cos(ang), math.sin(ang)) * dist
            pos.x = max(60.0, min(S.WORLD_WIDTH - 60.0, pos.x))
            pos.y = max(60.0, min(S.WORLD_HEIGHT - 60.0, pos.y))
            if not world_map.blocked(pos, radius):
                return pos
        return None

File: web/src/game/spawner.ts
// src/game/spawner.ts
// ZombieSpawner: positions + weighted type selection.

import { SPAWN_MAX_DIST, SPAWN_MIN_DIST, WORLD_HEIGHT, WORLD_WIDTH } from "./settings";
import { mulberry32, type Rng } from "../lib/rng";
import { createZombie, type Zombie } from "./zombie";
import type { ZombieData } from "./data";
import type { GameMap } from "./map";
import type { Vec } from "./vec";

export class ZombieSpawner {
  rng: Rng = mulberry32(Math.floor(Math.random() * 2 ** 31));

  pickType(wave: number): string {
    const weights: Record<string, number> = { normal: 10 };
    if (wave >= 2) weights.fast = Math.min(6, 2 + wave * 0.5);
    if (wave >= 3) weights.tank = Math.min(5, 1 + (wave - 2) * 0.4);
    if (wave >= 4) {
      weights.exploder = Math.min(4, 1 + (wave - 3) * 0.35);
      weights.ranged = Math.min(4, 1 + (wave - 3) * 0.35);
    }
    const total = Object.values(weights).reduce((a, b) => a + b, 0);
    let roll = this.rng.next() * total;
    for (const [k, w] of Object.entries(weights)) {
      roll -= w;
      if (roll <= 0) return k;
    }
    return "normal";
  }

  spawnPosition(playerPos: Vec, world: GameMap, radius = 24): Vec | null {
    for (let i = 0; i < 40; i++) {
      const ang = this.rng.next() * Math.PI * 2;
      const dist = SPAWN_MIN_DIST + this.rng.next() * (SPAWN_MAX_DIST - SPAWN_MIN_DIST);
      const pos: Vec = {
        x: Math.max(60, Math.min(WORLD_WIDTH - 60, playerPos.x + Math.cos(ang) * dist)),
        y: Math.max(60, Math.min(WORLD_HEIGHT - 60, playerPos.y + Math.sin(ang) * dist)),
      };
      if (!world.blocked(pos, radius)) return pos;
    }
    return null;
  }

  makeZombie(
    kind: string,
    pos: Vec,
    data: Record<string, ZombieData>,
    wave: number,
    night: number,
  ): Zombie {
    // wave scaling matches game.py: hp_growth, speed_growth, damage_growth
    const waveIdx = Math.max(0, wave - 1);
    const hpMult = 1 + 0.08 * waveIdx;
    const speedMult = (1 + 0.02 * waveIdx) * (1 + 0.3 * night);
    const dmgMult = (1 + 0.04 * waveIdx) * (1 + 0.25 * night);
    return createZombie(kind, pos, data, hpMult, speedMult, dmgMult);
  }
}
